### ISSUE TITLE
Denote mutates passed-in lists when sorting

### ISSUE THREAD

**Issue by mentalisttraceur** [protesilaos/denote#270]
Here's a fun one: given that `some-keywords` is a persistent variable (global or buffer-local, for example), what's the bug in this line?

```lisp
(denote title (cons special-keyword some-keywords))
```

If it's not immediately obvious, I hope that makes my point.

The bug is that if keyword sorting is turned on, `some-keywords` might get reordered, or even changed so that `special-keyword` ends up inside `some-keywords`, and/or `some-keywords` loses one or more keywords.

For example, try `(defvar some-keywords '("aa"))`. Then see what the contents of `some-keywords` become after a call to `(denote "foo" (cons "qq" some-keywords))`.

The caller-side fix is to copy the keyword list before passing it to `denote` if you don't want it mutated, but I would argue that the benefit of users not needing to remember that is much higher than any benefits of avoiding a list copy, since this is a user-facing UX function which is called realistically <1000 times a day with typically <10 keywords, not something called in a hot loop on millions of keywords, so it's probably better to solve it internally.

So I suggest either

1. changing `denote-keywords-sort` to do `(sort (copy-sequence keywords) ...)`, or

2. at least making `denote` copy the keyword list before passing it to `denote-keywords-sort`, or if you're really attached to mutating in-place sort

3. at least warn that the list might be mutated in the docstring `denote-keywords-sort` and every function that calls it. Could reuse wording that `sort` uses in its own docstring - "KEYWORDS is modified by side effects." Or maybe better, using more conventional modern language: "KEYWORDS is sorted in-place".

But I'd probably go with option one. It's considered best practice in software nowadays to not mutate data unless you have to. Fancy optimizer passes can turn copies and immutable semantics into in-place mutations fairly easily. Meanwhile, humans trying to get things done struggle to remember and stay vigilant about data-mutating semantics, which are transitive and can reach arbitrarily far across the program - a list can be passed through many functions and sit around in variables for a long time, and anything that touches it could break some invariant by mutating it.

This is from a real use-case. Earlier this evening I had some code unexpectedly misbehave when I was trying to lean on it. I've built a basic to-do/task system on top of Denote+Dired that I rely on throughout the day, where each task is a note file with a special keyword. I was trying to create new tasks with specific keywords, and any time I created more than one new task in the same task list without recreating that list's buffer, all but the first were missing one of the keywords (because after the sort, the special keyword had replaced one of the other keywords in the reused buffer-local list). The sort function was almost the last place I thought to look, because so many other Emacs functions are non-mutating by default, and have the mutating version tucked away under a weird name (c.f. `append` vs `nconc`, `take` vs `ntake`, and so on - I was *shocked*, flabbergasted, astounded, offended, [...], when I discovered that the built-in `sort` mutates in-place instead of creating a copy - I would've thought there was a separate function named something like `nsort` for that).

TL;DR: plz patch:

```diff
 (defun denote-keywords-sort (keywords)
   "Sort KEYWORDS if `denote-sort-keywords' is non-nil.
 KEYWORDS is a list of strings, per `denote-keywords-prompt'."
   (if denote-sort-keywords
-      (sort keywords #'string-collate-lessp)
+      (sort (copy-sequence keywords) #'string-collate-lessp)
     keywords))
```

**Comment by mentalisttraceur** [comment]
In the meantime, here's the advice I'm using to inject a fix locally, in case anyone else is running into the same issue:

```lisp
    (defun fixed-denote-keywords-sort (arguments)
        (when denote-sort-keywords
            (setcar arguments (copy-sequence (car arguments))))
        arguments)
    (advice-add 'denote-keywords-sort :filter-args 'fixed-denote-keywords-sort)
```

**Comment by protesilaos** [comment]
> From: mentalisttraceur ***@***.***>
> Date: Tue,  5 Mar 2024 22:32:29 -0800
>
> [... 24 lines elided]
>
> TL;DR: plz patch:
>
> ```diff
>  (defun denote-keywords-sort (keywords)
>    "Sort KEYWORDS if `denote-sort-keywords' is non-nil.
>  KEYWORDS is a list of strings, per `denote-keywords-prompt'."
>    (if denote-sort-keywords
> -      (sort keywords #'string-collate-lessp)
> +      (sort (copy-sequence keywords) #'string-collate-lessp)
>      keywords))
> ```

Thank you! Do you want to send this as a patch or pull request? Just so
your contribution is recorded in Git.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by mentalisttraceur** [comment]
Thanks for the offer but credit for such a small change doesn't matter to me and it seems more mutually efficient to not bother.

**Comment by protesilaos** [comment]
Thank you!
